subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java

```diff
@@ -98,27 +101,29 @@ private synchronized Spec<FileTreeElement> getDefaultExcludeSpec(CaseSensitivity
         return defaultExcludeSpecCache.get(caseSensitivity);
     }
 
-    private boolean invalidChangeOfExcludes(String[] defaultExcludes) {
-        return !Arrays.equals(previousDefaultExcludes, defaultExcludes);
+    private boolean invalidChangeOfExcludes(Set<String> defaultExcludes) {
+        return !previousDefaultExcludes.equals(defaultExcludes);
     }
 
-    private void failOnChangedDefaultExcludes(String[] excludesFromSettings, String[] newDefaultExcludes) {
-        List<String> sortedExcludesFromSettings = Arrays.asList(excludesFromSettings);
+    private void failOnChangedDefaultExcludes(Set<String> excludesFromSettings, Set<String> newDefaultExcludes) {
+        List<String> sortedExcludesFromSettings = new ArrayList<String>(excludesFromSettings);
         sortedExcludesFromSettings.sort(Comparator.naturalOrder());
-        List<String> sortedNewExcludes = Arrays.asList(newDefaultExcludes);
+        List<String> sortedNewExcludes = new ArrayList<String>(newDefaultExcludes);
         sortedNewExcludes.sort(Comparator.naturalOrder());
         throw new InvalidUserCodeException(String.format("Cannot change default excludes during the build. They were changed from %s to %s. Configure default excludes in the settings script instead.",  sortedExcludesFromSettings, sortedNewExcludes));
     }
 
+    @SuppressWarnings("unchecked")
     public synchronized void setDefaultExcludesFromSettings(String[] excludesFromSettings) {
-        if (!Arrays.equals(previousDefaultExcludes, excludesFromSettings)) {
-            updateDefaultExcludeSpecCache(excludesFromSettings);
+        Set<String> excludesFromSettingsSet = new HashSet(Arrays.asList(excludesFromSettings));
+        if (!previousDefaultExcludes.equals(excludesFromSettingsSet)) {
+            updateDefaultExcludeSpecCache(excludesFromSettingsSet);
         }
     }
 
-    private void updateDefaultExcludeSpecCache(String[] defaultExcludes) {
+    private void updateDefaultExcludeSpecCache(Set<String> defaultExcludes) {
         previousDefaultExcludes = defaultExcludes;
-        List<String> patterns = Arrays.asList(defaultExcludes);
+        List<String> patterns = new ArrayList<String>(defaultExcludes);
```

Comment:
❌  `createSpec` accepts a `Collection`, so there is no need to cast the set to a list here.